Native support for displaying, summarising EO3

jeremyh opened the pull request:
- Display the dataset geometries for EO3 datasets (instead of just a max/min box).
- Support regions (individual, searchable grid cells with counts and statistics) for all products with a `region_code` (this includes not just EO3, but some older products)
- Update the document renderer for eo3 fields
- Add a sampler that finds product metadata that’s not included in the product definition. These are displayed with a squiggly underline.
- Move some of the product information to a separate product page.

Linked to #133

This includes a DB schema change, so users will need to rerun `cubedash-gen —init` to update their schemas.

jeremyh committed:
Add EO3 fields to doc renderer

For #133

jeremyh committed:
Add initial EO3 extent support

#133

jeremyh committed:
Test CRS, Geometry selection for EO3

jeremyh committed:
Add eo3 (and eo_plus) region code population

For #133

jeremyh committed:
Use ARD as EO3 testcase

It has better geometry.

jeremyh committed:
Initial, slow, EO3 region support

For #133

- Rendering needs optimisation
- It's only showing generic labels on the UI
- Probably need schema migration for the new region_codes?

jeremyh committed:
Remove grid chopping, cache it

Everything will now use the Grid table instead
of doing it live.

The caching here is susceptible to #131, but that
will be handled separately.

jeremyh committed:
Add a sampler for fixed metadata fields

Will be useful to put on UI soon

jeremyh committed:
Record fixed_fields in product table

jeremyh committed:
Run simplify on unioned region polygons

jeremyh committed:
Test updates for storing fixed fields on products

jeremyh committed:
Improve query plan of product fixed-field scan

jeremyh committed:
Add schema migration for new fields

jeremyh committed:
Remove the old region geometry calc

jeremyh committed:
Correctly take dataset sample

jeremyh committed:
Simplify some README explanations

jeremyh committed:
Add Synthesize geometry for telemetry

jeremyh commented:
Opening this now for more visibility. It’s been hidden in my own branch for a few weeks.

jeremyh committed:
Friendlier error: mismatched yaml types in products

jeremyh committed:
Show inferred fixed fields for products

jeremyh committed:
Cleaner display of a dataset's region

jeremyh committed:
Fix false positives for type mismatches

(test failure)

jeremyh committed:
Move product info to product page

A much more focused Overview page.

The product page does need more layout work done, though.

jeremyh committed:
Product-page layout consistency

jeremyh committed:
Add title-text to clarify null values

jeremyh committed:
Fix tests not initialising postgis

jeremyh committed:
Avoid text conversion of geoms

Haven't benchmarked if it's faster... but it's unnecessary.

jeremyh committed:
Update, expand tests for new region calc

jeremyh committed:
Remove wofs-summary regions assert

No region data is added by the test: it previously
worked because regions were drawn rather than
calculated from the real datasets.

jeremyh committed:
Uninitialised pages shouldn't show a region count

jeremyh committed:
Expect telemetry datasets in stac results now

jeremyh committed:
Expect multiple h2 tags in metadata type page

jeremyh committed:
Product names should link to product page

jeremyh committed:
Style data types as badges

jeremyh committed:
Seq test shouldn't care about product count

It failed with the recent test data additions.

We only care that the sequence isn't *incremented*, we don't mind
its exact value.

jeremyh committed:
Fix sequence-inc test not using the test db (!)

The exit_code wasn't being checked, nor the output, so it quietly succeeded
despite the generate call failing (and the sequence was indeed not being
incremented on the test db, so the test passed!).

jeremyh committed:
fixed-field update for new eo3 products

Extra fields were added in dea repo.

jeremyh committed:
Fix overview page for products with no datasets

jeremyh committed:
Update missing-region tooltip for new functionality

jeremyh committed:
Expand product page links, consistent style

jeremyh committed:
Clearer description of regions

jeremyh committed:
Allow areas with 0 regions (fix test)

jeremyh committed:
metadata-type page layout consistency

jeremyh commented:
This does not include all EO3 features we'd like (especially: stac api improvements). 

But it adds useful functionality, the tests now pass, and the PR is already too large.

(I'll be making a few more smaller PRs soon, so preferably no release immediately)

alexgleith commented on README.md:
```diff
@@ -10,7 +10,7 @@
 These directions are for running from a local folder in development. But it will run from any typical Python WSGI server. 
 
-Firstly, install Data Cube. Use of a [Data Cube conda environment](https://datacube-core.readthedocs.io/en/latest/ops/conda.html)
+Firstly, install Open Data Cube. Use of a [Data Cube conda environment](https://datacube-core.readthedocs.io/en/latest/ops/conda.html)
```
```suggestion
Firstly, install the Open Data Cube. Use of a [Data Cube conda environment](https://datacube-core.readthedocs.io/en/latest/ops/conda.html)
```

whatnick commented on cubedash/_utils.py:
```diff
@@ -415,6 +437,19 @@ def dataset_shape(ds: Dataset) -> Tuple[Optional[Polygon], bool]:
     return geom, True
 
 
+def to_wrs_84(geometry: BaseGeometry, crs: Union[int, str, dict, pyproj.CRS]):
```
Use `to_wgs_84`

alexgleith replied:
Tish pointed this out as it should be `wgs`

whatnick commented on cubedash/summary/_schema.py:
```diff
@@ -170,6 +176,16 @@
+REGION = Table(
+    "mv_region",
+    _REF_TABLE_METADATA,
+    Column("dataset_type_ref", SmallInteger, nullable=False),
```
Not to lump in many changes. Shall we bump primary key data type ?

jeremyh replied:
The corresponding upstream ODC-core primary key that this links to is a SmallInt, so I'm hesitant to deviate from it. We just need to avoid making bad upsert statements again...

alexgleith commented on cubedash/_utils.py:
```diff
@@ -415,6 +437,19 @@ def dataset_shape(ds: Dataset) -> Tuple[Optional[Polygon], bool]:
+def to_wrs_84(geometry: BaseGeometry, crs: Union[int, str, dict, pyproj.CRS]):
+    origin = pyproj.Proj(init=crs)
+    dest = pyproj.Proj(init="epsg:4326")
+    tranform_wrs84 = functools.partial(pyproj.transform, origin, dest)
```
There are datacube functions that do this, which are faster, somehow!

See: https://github.com/opendatacube/odc-tools/blob/develop/libs/index/odc/index/stac.py#L76

jeremyh replied:
Thanks -- that api is much cleaner than doing it ourselves.

Added.

alexgleith commented on cubedash/_utils.py:
```diff
@@ -415,6 +437,19 @@ def dataset_shape(ds: Dataset) -> Tuple[Optional[Polygon], bool]:
+    tranform_wrs84 = functools.partial(pyproj.transform, origin, dest)
+    new_geometry = transform(tranform_wrs84, geometry)
+    # Unwrap coordinates, if necessary
+    new_geometry = test_wrap_coordinates(new_geometry)
```
@Kirill888  implemented something for anti-meridean handling too, which is probably better than my effort.

alexgleith reviewed (commented):
Impressive piece of work! I haven't gone through it all, but managed to make a couple of suggestions.

whatnick reviewed (changes requested):
Small typo and a comment. Looks clean otherwise.

jeremyh committed:
readme language grammar

Co-authored-by: Alex Leith <alexgleith+github@gmail.com>

jeremyh committed:
Replace dateline-wrapping with ODC's

jeremyh committed:
Update antimeridian test case for proper cutting

The old one falsely failed with ODC's outputs because of
the proper cutting (Rather than going above 180).

Verified via eyeball, by plotting the two outputs.

jeremyh committed:
Use Postgis for mapping SRIDs to CRSes

Fixes test.

jeremyh committed:
Fix the frequent 'wrs' typo

jeremyh committed:
Update expected metadata type page header

jeremyh commented:
Moved to ODC's geometry functions -- which involved replacing the test data with more accurate outputs. I checked it by eyeballing the two geometries on a map.
